### PR TITLE
chore(rust): bump MSRV to 1.71.0

### PR DESCRIPTION
**Pull request by thoughtpolice**
Summary: Let's be more aggressive about tracking the latest stable Rust release. There's little benefit to being conservative so early on, especially when no users seem to have faced any issue with upgrading, or strictly required an old Rust version.

Right now, just lagging Rust by 1 major release probably seems fine. We're targeting 1.71.0 to get ahead of the curve, since 1.72.0 is will likely be out by the next `jj` release.

This also adds a `rust-toolchain` file to track that version as well. Why? See some of the discussion in #1836. In short, having the user on the specified minimum toolchain does help avoid some SNAFUs where they might use newer features. But that was only in the Nix Flake. Adding `rust-toolchain` explicitly means that `rustup` and other components will use this same version, too. That should help achieve some level of consistency.

Finally, this tweaks flake.nix to read the needed rust-version from the `rust- toolchain` file, to remove a little repetition.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)

### CODE REVIEW

**Comment by thoughtpolice** [comment]
It looks like the toolchain file is interfering with the Clippy and rustfmt actions. I'll have to poke it with a stick I think.

File: docs/contributing.md [hunk 1]
```diff
@@ -89,7 +89,7 @@ During development (adapt according to your preference):
 
     cargo watch --ignore '.jj/**' -s \
       'cargo clippy --workspace --all-targets \
-       && cargo +1.64 check --workspace --all-targets'
+       && cargo +1.70 check --workspace --all-targets'
```
Comment (necauqua) [inline docs/contributing.md:92]: shouldn't having explicit `+1.70` be not necessary here as cargo would read the toolchain file?

**Review by necauqua — CHANGES_REQUESTED** [review]

File: docs/contributing.md (hunk 1 above)
Comment (necauqua) [reply docs/contributing.md:92]: git review ui ate a large part of my comment (because I edited it), lol

I said that it's interesting that besides docs and CI (which can potentially read the toolchain file like Nix does) there are two sources of truth, the toolchain file and Cargo.toml, and cargo probably would read the latter as the toolchain file is more of a rustup thing iirc

**Comment by ilyagr** [comment]
I do think we should become more conservative about our MSRV later, but perhaps you're right that it's not worth it to worry about that now. Perhaps by the time `jj` is stable enough to be packaged, [Debian](https://packages.debian.org/search?keywords=rustc) will get up to 1.70.

**Comment by ilyagr** [comment]
Also, should we wait until *after* releasing 0.8.0 to merge this? There seems to be no cost to releasing another version with support for older Rust.

**Comment by thoughtpolice** [comment]
> Also, should we wait until _after_ releasing 0.8.0 to merge this?

Sure, sounds good to me. I can just rebase the changelog entry. 1.71.0 should come out on Thursday as well, so by that point we'll already be 1 release behind, which I think is good cadence.

**Comment by thoughtpolice** [comment]
This also now includes a small change to remove the `jj-lib` build script and the `nightly_shims` module, since it's no longer necessary.

**Comment by thoughtpolice** [comment]
@ilyagr I also removed your changes introduced in 738f99ddf, since I don't think those are necessary any longer either.

**Comment by ilyagr** [comment]
Thanks! I think every version of Rust now avoids that problem, see https://github.com/martinvonz/jj/pull/1686#issuecomment-1628312857.

Could you double-check that it's not a problem with either `cargo +nightly` or `cargo +beta`? It might trip up some people who use the `beta` toolchain until they run `rustup update`.

**Review by necauqua — APPROVED** [review]

File: lib/tests/test_revset.rs [hunk 2]
```diff
@@ -12,11 +12,6 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// TODO: Remove after https://github.com/frondeus/test-case/issues/122 is resolved
```
Comment (ilyagr) [inline lib/tests/test_revset.rs:15]: Re commit message: To be clear, only the `allow(unknown_lints)` part was there for clippy < 1.70. The actual problem affected nighlites and `cargo +beta clippy` until a week ago.

But I think https://github.com/frondeus/test-case/issues/122 is fixed now.

File: docs/contributing.md [hunk 3]
```diff
@@ -1,6 +1,5 @@
 # How to Contribute
 
-
```
Comment (martinvonz) [inline docs/contributing.md:3]: nit: the formatting changes would ideally go into its own commit (not sure if you even intended them)

Comment (thoughtpolice) [reply docs/contributing.md:3]: Ugh, vscode autoformatting again. I'll try to rip this out.

File: rust-toolchain.toml [hunk 4]
```diff
@@ -0,0 +1,2 @@
+[toolchain]
+channel = "1.71.0"
```
Comment (martinvonz) [inline rust-toolchain.toml:2]: Does this mean that a user with a default toolchain version of 1.72.0 (once that's out) will still run the build with 1.71.0? It seemed that way when I tried it just now by adding a `rust-toolchain.toml` file with `channel = "1.60.0"`. It failed with this:
```
error: package `toml_datetime v0.6.3` cannot be built because it requires rustc 1.64.0 or newer, while the currently active rustc version is 1.60.0
```

Comment (thoughtpolice) [reply rust-toolchain.toml:2]: Yeah, if you're using `rustup`, it should automatically download and setup the correct version of the Rust compiler. Older versions won't work (as in your example), but newer versions should be fine. They'll just use that specific version of `rustc` when working in the repo.

Comment (martinvonz) [reply rust-toolchain.toml:2]: What I mean is that this seems to use version to 1.71 even if the user's default toolchain is 1.72. Is that what we want?

Comment (thoughtpolice) [reply rust-toolchain.toml]: Update: we discussed this on Discord, and I think the answer is "yes, this is what we want." Basically, we're going to invert the situation now where people use whatever Rust toolchain they have, and CI tests MSRV, and go in reverse: everyone develops with the MSRV, and the CI checks every change against the latest stable. The intent is to mainly catch diagnostics, clippy warnings, etc.

Comment (thoughtpolice) [reply rust-toolchain.toml]: I've also split out the `rust-toolchain.toml` change anyway, so we'll handle it in an upcoming idff.

File: docs/contributing.md [hunk 5]
```diff
@@ -116,9 +118,10 @@ These are listed roughly in order of decreasing importance.
    toolchain and use `cargo +nightly fmt`.
 
 3. Your code will be rejected if it cannot be compiled with the minimal
-   supported version of Rust ("MSRV"). This version is listed as
-   `rust-version` in [`Cargo.toml`](../Cargo.toml); it is 1.64 as of this
-   writing.
+   supported version of Rust ("MSRV"). Currently, `jj` follows a rather
+   casual MSRV policy: "The current `rustc` stable version, minus one."
+   This version is always listed in
+   [`rust-toolchain.toml`](../rust-toolchain.toml).
```
Comment (martinvonz) [inline docs/contributing.md:124]: This should also be moved to that other PR now. The commit message's mention of rust-toolchain should also be (re)moved
